### PR TITLE
wb32-dfu-updater_cli 0.0.3 (new formula)

### PR DESCRIPTION
**Pull request by itarze**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

File: Formula/wb32-dfu-updater_cli.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
+class DfuUtil < Formula
+  desc "USB programmer"
+  homepage "https://github.com/WestberryTech/wb32-dfu-updater"
+  url "https://github.com/WestberryTech/wb32-dfu-updater/archive/refs/tags/v0.0.1.tar.gz"
+  sha256 "33e7aeb616bdd8e4f4b92de9c7a15d35e0188683e18f2d394728e2c74798d900"
+  license "Apache-2.0"
+  revision 1
```
Comment (SMillerDev) [inline Formula/wb32-dfu-updater_cli.rb:7]: This should be left out by default
```suggestion

```

File: Formula/wb32-dfu-updater_cli.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
+class DfuUtil < Formula
```
Comment (SMillerDev) [inline Formula/wb32-dfu-updater_cli.rb:1]:
```suggestion
class Wb32DfuUpdaterCli < Formula

```

File: Formula/wb32-dfu-updater_cli.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    mkdir "build" do
+      system "cmake", "../source/wb32-dfu-updater_cli",
+                      "-CMAKE_INSTALL_PREFIX=#{rpath}"
+      system "make"
+      system "make", "install"
+    end
+  end
+
+  test do
+    system bin/"wb32-dfu-updater_cli", "-V"
```
[the remark this answers is not in the document]
Comment (wb-Joy) [reply Formula/wb32-dfu-updater_cli.rb:22]: Testing this tool requires hardware support, without hardware can only test the version output.

File: Formula/wb32-dfu-updater_cli.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+  depends_on "cmake" => :build
+  depends_on "libusb"
+
+  def install
+    mkdir "build" do
+      system "cmake", "../source/wb32-dfu-updater_cli",
+                      *std_cmake_args
+      system "make", "install"
+    end
```
Comment (carlocab) [inline Formula/wb32-dfu-updater_cli.rb:17]:
```suggestion
    system "cmake", "-S", "source/wb32-dfu-updater_cli", "-B", "build", *std_cmake_args
    system "cmake", "--build", "build"
    system "cmake", "--install", "build"
```

File: Formula/wb32-dfu-updater_cli.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (carlocab) [reply Formula/wb32-dfu-updater_cli.rb]: Can we try to use it with non-existent hardware and check we get the expected error/failure message?

[the remark this answers is not in the document]
Comment (wb-Joy) [reply Formula/wb32-dfu-updater_cli.rb]: An MCU with USB is required and specific firmware is also required.

[the remark this answers is not in the document]
Comment (wb-Joy) [reply Formula/wb32-dfu-updater_cli.rb]: This software is used for keyboard firmware upgrade.

[the remark this answers is not in the document]
Comment (SMillerDev) [reply Formula/wb32-dfu-updater_cli.rb]: And if you give it data that it doesn't work with? It gives an error right?

[the remark this answers is not in the document]
Comment (wb-Joy) [reply Formula/wb32-dfu-updater_cli.rb]: It will give a hint of the correct parameters.

[the remark this answers is not in the document]
Comment (wb-Joy) [reply Formula/wb32-dfu-updater_cli.rb]: I am an MCU developer, this is my first time developing desktop-level software, I don't understand why there are errors in the test.

**Comment by Bo98** [comment]
The default permissions seem unusual:

```
---x------ 1 bo admin 29324 Dec 11 01:51 /usr/local/Cellar/wb32-dfu-updater_cli/0.0.2/bin/wb32-dfu-updater_cli
```

`brew` can't even read the file which is why there's errors.

**Comment by wb-Joy** [comment]
All the checks have passed, so can the PR be merged?

**Review by SMillerDev — CHANGES_REQUESTED** [review]
The test still needs to be improved.

It's also self-submitted software, which we generally advise against.

**Comment by wb-Joy** [comment]
> 
I don't quite understand, I have seen that the test of dfu-util is the same as mine, only printing their version number.
image

**Comment by SMillerDev** [comment]
> I don't quite understand, I have seen that the test of dfu-util is the same as mine, only printing their version number.

Just because old code is bad does not mean that new code must be bad too.

**Comment by wb-Joy** [comment]
> > I don't quite understand, I have seen that the test of dfu-util is the same as mine, only printing their version number.
> 
> Just because old code is bad does not mean that new code must be bad too.

I added an entry that shows an item of devices. It can only do so much without a hardware connection.

**Comment by wb-Joy** [comment]
When connecting a hardware device, you can run the -D command to download the firmware or run the -U command to read the firmware. The Windows version msys2 is passed.

File: Formula/wb32-dfu-updater_cli.rb [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+  depends_on "cmake" => :build
+  depends_on "libusb"
+
+  def install
+    args = std_cmake_args
+
+    libusb = Formula["libusb"]
+    args << "-DLIBUSB_INCLUDE_DIRS=#{libusb.opt_include}/libusb-1.0"
+    args << "-DLIBUSB_LIBRARIES=#{libusb.opt_lib}/#{shared_library("libusb-1.0")}"
```
Comment (carlocab) [inline Formula/wb32-dfu-updater_cli.rb:17]:
```suggestion
    args << "-DLIBUSB_LIBRARIES=#{libusb.opt_lib/shared_library("libusb-1.0")}"
```
Interpolating twice isn't necessary here.

File: Formula/wb32-dfu-updater_cli.rb [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+    mkdir "build" do
+      system "cmake", "..", *args
+      system "make", "install"
+    end
```
Comment (carlocab) [inline Formula/wb32-dfu-updater_cli.rb:22]:
```suggestion
    system "cmake", "-S", ".", "-B", "build", *args
    system "cmake", "--build", "build"
    system "cmake", "--install", "build"
```
Let's follow the template from `brew create --cmake`.

File: Formula/wb32-dfu-updater_cli.rb [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  end
+
+  test do
+    printing_message = "wb32-dfu-updater ver: #{version}\n \
+                        libusb version 1.0.24 (11584)\n \
+                        ----------------------------------------\n"
```
Comment (carlocab) [inline Formula/wb32-dfu-updater_cli.rb:28]:
```suggestion
    printing_message = <<~EOS
      wb32-dfu-updater ver: #{version}
      libusb version #{Formula["libusb"].version} (11584)
      ----------------------------------------
    EOS
```
Not quite sure where the `11584` comes from, but I don't think that's something we want to keep static here.

File: Formula/wb32-dfu-updater_cli.rb [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    system "cmake", "--install", "build"
+  end
+
+  test do
+    printing_message = <<~EOS
+      wb32-dfu-updater ver: #{version}
+      libusb version #{Formula["libusb"].version} (11584)
+      ----------------------------------------
+    EOS
+    assert_match printing_message, shell_output(bin/"wb32-dfu-updater_cli --list")
```
Comment (carlocab) [inline Formula/wb32-dfu-updater_cli.rb:30]:
```suggestion
    assert_match printing_message, shell_output("#{bin}/wb32-dfu-updater_cli --list")
```

File: Formula/wb32-dfu-updater_cli.rb [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+    system "cmake", "-S", ".", "-B", "build", *args
+    system "cmake", "--build", "build"
+    system "cmake", "--install", "build"
+  end
+
+  test do
+    printing_message = <<~EOS
+      wb32-dfu-updater ver: #{version}
+      libusb version #{Formula["libusb"].version} (11584)
```
Comment (carlocab) [inline Formula/wb32-dfu-updater_cli.rb:27]: Where does `11584` come from?

**Review by carlocab — COMMENTED** [review]
Apologies for the delay, and thanks so much for your work here.

Your repository doesn't quite meet our notability requirements, and we usually don't like self-submitted stuff because that's a poor indicator of whether any of our users will actually use this formula.

If one of your users resubmits this when your repository is slightly more notable (e.g. stars/forks, or is packaged by some other package managers/distros), we'd be happy to reconsider.

File: Formula/wb32-dfu-updater_cli.rb (hunk 9 above)
Comment (wb-Joy) [reply Formula/wb32-dfu-updater_cli.rb:27]: > Where does `11584` come from?

version nano.

**Pull request closed without merging**